File: web/bench.mjs
import os from "node:os";
import fs from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";

export const PRESETS = {
  spec: {
    title: "Issue #53 BLE Concurrency Spec",
    prompt: `# Kokolog Monitor: BLE Auto-Connect & Fault Recovery Specification
Analyze the following acceptance criteria for Issue #53:
1. The runtime monitors advertisements for all registered devices and connects up to four simultaneously.
2. ECG, acceleration, and rPos reception is independent per device.
3. One failing device does not block the other three.
4. GATT 133 closes resources before reconnection, and duplicate concurrent attempts do not occur.
5. A normal reconnect attempt starts within two seconds.
6. Bluetooth-off, scan-restricted, permission, and terminal states provide distinct guidance.

Task:
- Identify the 3 critical race conditions in this 4-slot BLE connection and reconnect state machine.
- Provide a concise architectural recommendation for GATT 133 resource closure.
`
  },
  bug: {
    title: "Dart State Reducer Missing Case",
    prompt: `// Identify the bug and fix the missing state mapping in this Dart reducer snippet:
MonitorAppState reduceAppState({
  required HrmRuntimeState snapshot,
  required bool isSnapshotFresh,
  required TransportHealth? transportHealth,
  required bool isConfigured,
}) {
  if (!isConfigured) return MonitorAppState.unconfigured();
  return switch (snapshot.availability) {
    HrmRuntimeAvailability.ready => _reduceReady(snapshot),
    HrmRuntimeAvailability.bluetoothOff => MonitorAppState.actionRequired(ActionRequiredKind.bluetoothOff),
    HrmRuntimeAvailability.permissionRequired => MonitorAppState.actionRequired(ActionRequiredKind.permissionRequired),
    HrmRuntimeAvailability.unsupported => MonitorAppState.actionRequired(ActionRequiredKind.unsupportedDevice),
    // Defect: backgroundRestricted / scanRestricted fall through to stopped(runtimeUnavailable)
    _ => MonitorAppState.stopped(reason: StoppedReason.runtimeUnavailable),
  };
}
// Task: Explain the defect and output the corrected Dart code block.
`
  },
  arch: {
    title: "Loop Lifecycle & Role Handoff",
    prompt: `Explain the handoff contract in an autonomous coding loop between:
1. Implementer (maker)
2. Reviewer / Assurance (checker)
3. Gardener (ticket backfill)
4. Decision Desk (human gates)

State plainly what an implementer must do when encountering a stale test or an adjacent bug.
`
  }
};

async function loadOpencodeConfig() {
  const candidates = [
    path.join(os.homedir(), ".config/opencode/opencode.json"),
    path.join(os.homedir(), ".opencode/opencode.json")
  ];
  for (const file of candidates) {
    try {
      const text = await fs.readFile(file, "utf8");
      return JSON.parse(text);
    } catch {}
  }
  return {};
}

export async function resolveBenchConfig() {
  const cfg = await loadOpencodeConfig();
  const p10 = cfg.provider?.["10router"] || {};
  const models = [];
  for (const [id, m] of Object.entries(p10.models || {})) {
    models.push({
      id: `10router/${id}`,
      name: id,
      variants: Object.keys(m.variants || {})
    });
  }
  return {
    engines: [
      { id: "opencode", name: "OpenCode CLI (Real Agent Runtime)" },
      { id: "gateway", name: "Direct Gateway (Raw API Baseline)" }
    ],
    models,
    presets: PRESETS
  };
}

export async function resolveGatewayConfig() {
  const cfg = await loadOpencodeConfig();
  const p10 = cfg.provider?.["10router"] || {};
  const baseURL = (p10.options?.baseURL || "https://ai.zcloud.my/v1").replace(/\/$/, "");
  const apiKey = p10.options?.apiKey || process.env.OPENCODE_API_KEY || "";
  const models = Object.keys(p10.models || {});
  return { baseURL, apiKey, models };
}

export async function handleBench(request, response) {
  let raw = "";
  for await (const chunk of request) {
    raw += chunk;
    if (raw.length > 65536) {
      response.writeHead(413, { "content-type": "application/json" });
      return response.end(JSON.stringify({ error: "Prompt too large (max 64KB)" }));
    }
  }

  let body;
  try {
    body = JSON.parse(raw);
  } catch {
    response.writeHead(400, { "content-type": "application/json" });
    return response.end(JSON.stringify({ error: "Malformed JSON" }));
  }

  const {
    engine = "opencode",
    model = "10router/combo-coding",
    variant = "high",
    thinkingLevel = "high",
    prompt = ""
  } = body;

  if (!prompt.trim()) {
    response.writeHead(400, { "content-type": "application/json" });
    return response.end(JSON.stringify({ error: "Missing prompt" }));
  }

  response.writeHead(200, {
    "content-type": "text/event-stream; charset=utf-8",
    "cache-control": "no-cache, no-transform",
    connection: "keep-alive"
  });

  const sendEvent = (event, data) => {
    response.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
  };

  if (engine === "opencode") {
    return runOpencodeBench({ model, variant, thinkingLevel, prompt, sendEvent, request, response });
  } else {
    return runGatewayBench({ model, thinkingLevel, prompt, sendEvent, response });
  }
}

async function runOpencodeBench({ model, variant, thinkingLevel, prompt, sendEvent, request, response }) {
  const tStart = Date.now();
  let tFirstStep = null;
  let tFirstContent = null;
  let tFirstChunk = null;

  let stepCount = 0;
  let toolsUsed = [];
  let totalReasoningTokens = 0;
  let totalOutputTokens = 0;
  let totalInputTokens = 0;
  let fullContent = "";

  const opencodeBin = process.env.OPENCODE_BIN || "/Users/zafar/.opencode/bin/opencode";
  const title = `bench-${Date.now().toString(36)}`;

  let modelArg = model;
  if (!modelArg.startsWith("10router/") && !modelArg.includes("/")) {
    modelArg = `10router/${modelArg}`;
  }

  const benchDir = "/tmp/bench-isolated";
  await fs.mkdir(benchDir, { recursive: true }).catch(() => {});

  const args = ["run", "--dir", benchDir, "--pure", "--model", modelArg];
  const chosenVariant = variant || (thinkingLevel !== "none" ? thinkingLevel : "");
  if (chosenVariant && chosenVariant !== "none") {
    args.push("--variant", chosenVariant);
  }
  args.push("--format", "json", "--title", title, "--", prompt);

  sendEvent("start", { engine: "opencode", model: modelArg, variant: chosenVariant, tStart });

  const env = {
    ...process.env,
    OPENCODE_DB: "/tmp/bench-opencode.db",
    OPENCODE_CONFIG_CONTENT: JSON.stringify({
      plugin: [],
      mcp: { "google-docs": { enabled: false } }
    }),
    OPENCODE_PERMISSION: JSON.stringify({
      edit: "deny",
      task: "deny",
      external_directory: "deny",
      bash: "deny"
    })
  };

  const child = spawn(opencodeBin, args, {
    cwd: benchDir,
    stdio: ["ignore", "pipe", "pipe"],
    env
  });

  let killed = false;
  const cleanupChild = () => {
    if (!killed) {
      killed = true;
      try {
        child.kill("SIGTERM");
        setTimeout(() => {
          try { child.kill("SIGKILL"); } catch {}
        }, 1000);
      } catch {}
    }
  };

  request.on("close", cleanupChild);

  let stdoutBuffer = "";

  child.stdout.on("data", (chunk) => {
    stdoutBuffer += chunk.toString("utf8");
    const lines = stdoutBuffer.split("\n");
    stdoutBuffer = lines.pop() || "";

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;

      let event;
      try {
        event = JSON.parse(trimmed);
      } catch {
        continue;
      }

      const now = Date.now();
      if (tFirstChunk === null) tFirstChunk = now;

      const type = event.type;
      const part = event.part || {};

      if (type === "step_start") {
        stepCount++;
        if (tFirstStep === null) {
          tFirstStep = now;
          const startupSec = (tFirstStep - tStart) / 1000;
          sendEvent("startup", { startup_sec: Number(startupSec.toFixed(3)) });
        }
        sendEvent("step_start", { step: stepCount });
      } else if (type === "text") {
        if (tFirstContent === null) {
          tFirstContent = now;
          const ttftSec = (tFirstContent - tStart) / 1000;
          sendEvent("ttft", { ttft_sec: Number(ttftSec.toFixed(3)) });
        }
        const txt = part.text || "";
        if (txt) {
          fullContent += txt;
          sendEvent("content", { text: txt });
        }
      } else if (type === "tool_use") {
        if (tFirstContent === null) {
          tFirstContent = now;
          const ttftSec = (tFirstContent - tStart) / 1000;
          sendEvent("ttft", { ttft_sec: Number(ttftSec.toFixed(3)) });
        }
        const toolName = part.tool || part.name || "unknown";
        toolsUsed.push(toolName);
        sendEvent("tool", { step: stepCount, tool: toolName, input: part.input });
      } else if (type === "step_finish") {
        const tokens = part.tokens || {};
        totalReasoningTokens += tokens.reasoning || 0;
        totalOutputTokens += tokens.output || 0;
        totalInputTokens = Math.max(totalInputTokens, tokens.input || 0);

        sendEvent("step_finish", {
          step: stepCount,
          tokens: {
            reasoning: tokens.reasoning || 0,
            output: tokens.output || 0,
            input: tokens.input || 0,
            total: tokens.total || 0
          }
        });
      }
    }
  });

  let stderrLog = "";
  child.stderr.on("data", (chunk) => {
    stderrLog += chunk.toString("utf8");
  });

  child.on("error", (err) => {
    sendEvent("error", { message: `Failed to spawn OpenCode: ${err.message}` });
    response.end();
  });

  child.on("close", (code) => {
    const tEnd = Date.now();
    const totalSec = Math.max(0.001, (tEnd - tStart) / 1000);
    const startupSec = tFirstStep ? Math.max(0, (tFirstStep - tStart) / 1000) : 0;
    const ttftSec = tFirstContent ? Math.max(0, (tFirstContent - tStart) / 1000) : totalSec;

    const totalGeneratedTokens = totalReasoningTokens + totalOutputTokens;
    const genSec = Math.max(0.001, totalSec - startupSec);
    const tps = totalGeneratedTokens / genSec;

    if (code !== 0 && !fullContent && !toolsUsed.length) {
      sendEvent("error", {
        message: `OpenCode exited with code ${code}. Stderr: ${stderrLog.slice(0, 500) || "none"}`
      });
      return response.end();
    }

    sendEvent("done", {
      engine: "opencode",
      model: modelArg,
      variant: chosenVariant,
      startup_sec: Number(startupSec.toFixed(3)),
      ttft_sec: Number(ttftSec.toFixed(3)),
      thinking_dur_sec: Number(Math.max(0, ttftSec - startupSec).toFixed(3)),
      thinking_tokens: totalReasoningTokens,
      output_dur_sec: Number(Math.max(0, totalSec - ttftSec).toFixed(3)),
      output_tokens: totalOutputTokens,
      input_tokens: totalInputTokens,
      total_dur_sec: Number(totalSec.toFixed(3)),
      total_tokens: totalGeneratedTokens,
      steps: stepCount,
      tools: toolsUsed,
      tps: Number(tps.toFixed(1)),
      exit_code: code
    });
    response.end();
  });
}

async function runGatewayBench({ model, thinkingLevel, prompt, sendEvent, response }) {
  const { baseURL, apiKey } = await resolveGatewayConfig();
  if (!apiKey) {
    sendEvent("error", { message: "No API key configured in ~/.config/opencode/opencode.json" });
    return response.end();
  }

  let targetModel = model.replace(/^10router\//, "");
  if (targetModel.startsWith("ag/gemini-") && ["low", "medium", "high"].includes(thinkingLevel)) {
    const parts = targetModel.split("-");
    if (["low", "medium", "high"].includes(parts[parts.length - 1])) {
      parts[parts.length - 1] = thinkingLevel;
      targetModel = parts.join("-");
    } else {
      targetModel = `${targetModel}-${thinkingLevel}`;
    }
  }

  const payload = {
    model: targetModel,
    messages: [{ role: "user", content: prompt }],
    stream: true,
    max_tokens: 1500
  };

  if (["low", "medium", "high"].includes(thinkingLevel)) {
    payload.reasoning_effort = thinkingLevel;
    payload.thinkingConfig = { thinkingLevel };
  }

  const tStart = Date.now();
  let tFirstToken = null;
  let tFirstContent = null;
  let tReasoningStart = null;
  let tReasoningEnd = null;

  let reasoningText = "";
  let contentText = "";
  let reportedReasoningTokens = null;
  let reportedOutputTokens = null;

  try {
    sendEvent("start", { engine: "gateway", model: targetModel, thinkingLevel, tStart });

    const upstream = await fetch(`${baseURL}/chat/completions`, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${apiKey}`,
        "Content-Type": "application/json",
        "User-Agent": "opencode/1.18.25"
      },
      body: JSON.stringify(payload)
    });

    if (!upstream.ok) {
      const errText = await upstream.text().catch(() => "");
      sendEvent("error", { message: `Gateway responded with ${upstream.status}: ${errText}` });
      return response.end();
    }

    const reader = upstream.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const rawLine of lines) {
        const line = rawLine.trim();
        if (!line || line.startsWith(":")) continue;
        if (line.startsWith("data: ")) {
          const dataStr = line.slice(6);
          if (dataStr === "[DONE]") break;

          let chunk;
          try {
            chunk = JSON.parse(dataStr);
          } catch {
            continue;
          }

          const now = Date.now();
          if (tFirstToken === null) {
            tFirstToken = now;
            sendEvent("ttft", { ttft_sec: Number(((tFirstToken - tStart) / 1000).toFixed(3)) });
          }

          const delta = chunk.choices?.[0]?.delta || {};
          const r = delta.reasoning_content || delta.thinking || "";
          const c = delta.content || "";

          if (r) {
            if (tReasoningStart === null) tReasoningStart = now;
            tReasoningEnd = now;
            reasoningText += r;
            sendEvent("reasoning", { text: r });
          }

          if (c) {
            if (tFirstContent === null) tFirstContent = now;
            contentText += c;
            sendEvent("content", { text: c });
          }

          const usage = chunk.usage;
          if (usage) {
            if (usage.completion_tokens_details?.reasoning_tokens !== undefined) {
              reportedReasoningTokens = usage.completion_tokens_details.reasoning_tokens;
            }
            if (usage.completion_tokens !== undefined) {
              reportedOutputTokens = usage.completion_tokens;
            }
          }
        }
      }
    }

    const tEnd = Date.now();
    const totalSec = Math.max(0.001, (tEnd - tStart) / 1000);
    const ttftSec = tFirstToken ? (tFirstToken - tStart) / 1000 : totalSec;

    const rTokens = reportedReasoningTokens ?? Math.max(Math.round(reasoningText.length / 4), reasoningText ? 1 : 0);
    const oTokens = reportedOutputTokens ?? Math.max(Math.round(contentText.length / 4), contentText ? 1 : 0);

    let thinkingDurSec = 0;
    if (reasoningText && tFirstContent) {
      thinkingDurSec = Math.max(0, (tFirstContent - tFirstToken) / 1000);
    } else if (reasoningText && tReasoningEnd && tReasoningStart) {
      thinkingDurSec = Math.max(0, (tReasoningEnd - tReasoningStart) / 1000);
    }

    const outputDurSec = Math.max(0, (tEnd - (tFirstContent || tFirstToken || tEnd)) / 1000);
    const totalTokens = rTokens + oTokens;
    const tps = totalTokens / totalSec;

    sendEvent("done", {
      engine: "gateway",
      model: targetModel,
      thinkingLevel,
      startup_sec: 0,
      ttft_sec: Number(ttftSec.toFixed(3)),
      thinking_dur_sec: Number(thinkingDurSec.toFixed(3)),
      thinking_tokens: rTokens,
      output_dur_sec: Number(outputDurSec.toFixed(3)),
      output_tokens: oTokens,
      total_dur_sec: Number(totalSec.toFixed(3)),
      total_tokens: totalTokens,
      steps: 1,
      tools: [],
      tps: Number(tps.toFixed(1))
    });
  } catch (err) {
    sendEvent("error", { message: err.message || "Streaming failed" });
  } finally {
    response.end();
  }
}

File: scripts/bench.py
#!/usr/bin/env python3
"""
scripts/bench.py — OpenCode Model + Thinking Speed & Latency Benchmark

Measures:
- OpenCode startup / bootstrap time
- TTFT (Time to First Token)
- Thinking / Reasoning duration & token count
- Output generation duration & token count
- Step count & Agent tool calls
- Throughput (tokens/sec)
- Total wall-clock turnaround

Supports:
- --engine opencode (default: runs real OpenCode agent runtime)
- --engine gateway (direct API baseline)
"""

import argparse
import json
import os
import subprocess
import sys
import time
import urllib.request
import urllib.error

PRESETS = {
    "spec": """# Kokolog Monitor: BLE Auto-Connect & Fault Recovery Specification
Analyze the following acceptance criteria for Issue #53:
1. The runtime monitors advertisements for all registered devices and connects up to four simultaneously.
2. ECG, acceleration, and rPos reception is independent per device.
3. One failing device does not block the other three.
4. GATT 133 closes resources before reconnection, and duplicate concurrent attempts do not occur.
5. A normal reconnect attempt starts within two seconds.
6. Bluetooth-off, scan-restricted, permission, and terminal states provide distinct guidance.

Task:
- Identify the 3 critical race conditions in this 4-slot BLE connection and reconnect state machine.
- Provide a concise architectural recommendation for GATT 133 resource closure.
""",
    "bug": """// Identify the bug and fix the missing state mapping in this Dart reducer snippet:
MonitorAppState reduceAppState({
  required HrmRuntimeState snapshot,
  required bool isSnapshotFresh,
  required TransportHealth? transportHealth,
  required bool isConfigured,
}) {
  if (!isConfigured) return MonitorAppState.unconfigured();
  return switch (snapshot.availability) {
    HrmRuntimeAvailability.ready => _reduceReady(snapshot),
    HrmRuntimeAvailability.bluetoothOff => MonitorAppState.actionRequired(ActionRequiredKind.bluetoothOff),
    HrmRuntimeAvailability.permissionRequired => MonitorAppState.actionRequired(ActionRequiredKind.permissionRequired),
    HrmRuntimeAvailability.unsupported => MonitorAppState.actionRequired(ActionRequiredKind.unsupportedDevice),
    // Defect: backgroundRestricted / scanRestricted fall through to stopped(runtimeUnavailable)
    _ => MonitorAppState.stopped(reason: StoppedReason.runtimeUnavailable),
  };
}
// Task: Explain the defect in 2 sentences and output the corrected Dart code block.
""",
    "arch": """Explain the handoff contract in an autonomous coding loop between:
1. Implementer (maker)
2. Reviewer / Assurance (checker)
3. Gardener (ticket backfill)
4. Decision Desk (human gates)

State plainly what an implementer must do when encountering a stale test or an adjacent bug.
"""
}

def load_opencode_config():
    paths = [
        os.path.expanduser("~/.config/opencode/opencode.json"),
        os.path.expanduser("~/.opencode/opencode.json"),
    ]
    for p in paths:
        if os.path.isfile(p):
            try:
                with open(p, "r", encoding="utf-8") as f:
                    return json.load(f)
            except Exception:
                pass
    return {}

def resolve_gateway():
    cfg = load_opencode_config()
    p10 = cfg.get("provider", {}).get("10router", {})
    opts = p10.get("options", {})
    base_url = opts.get("baseURL", "https://ai.zcloud.my/v1").rstrip("/")
    api_key = opts.get("apiKey", os.environ.get("OPENCODE_API_KEY", ""))
    return base_url, api_key

def run_opencode_bench(model, variant="high", prompt="hi", verbose=False):
    opencode_bin = os.environ.get("OPENCODE_BIN", "/Users/zafar/.opencode/bin/opencode")
    if not os.path.isfile(opencode_bin):
        print(f"Error: OpenCode binary not found at {opencode_bin}", file=sys.stderr)
        return None

    model_arg = model
    if not model_arg.startswith("10router/") and "/" not in model_arg:
        model_arg = f"10router/{model_arg}"

    os.makedirs("/tmp/bench-isolated", exist_ok=True)
    cmd = [
        opencode_bin,
        "run",
        "--dir", "/tmp/bench-isolated",
        "--pure",
        "--model", model_arg
    ]
    if variant and variant != "none":
        cmd.extend(["--variant", variant])
    cmd.extend(["--format", "json", "--title", f"bench-{int(time.time())}", "--", prompt])

    env = os.environ.copy()
    env["OPENCODE_DB"] = "/tmp/bench-opencode.db"
    env["OPENCODE_CONFIG_CONTENT"] = '{"plugin":[],"mcp":{"google-docs":{"enabled":false}}}'
    env["OPENCODE_PERMISSION"] = '{"edit":"deny","task":"deny","external_directory":"deny","bash":"deny"}'

    t_start = time.time()
    t_first_step = None
    t_first_content = None

    step_count = 0
    tools_used = []
    total_reasoning_tokens = 0
    total_output_tokens = 0
    full_content = []

    try:
        proc = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.DEVNULL,
            env=env,
            text=True
        )

        for raw_line in proc.stdout:
            line = raw_line.strip()
            if not line:
                continue

            try:
                event = json.loads(line)
            except Exception:
                continue

            now = time.time()
            ev_type = event.get("type")
            part = event.get("part", {})

            if ev_type == "step_start":
                step_count += 1
                if t_first_step is None:
                    t_first_step = now
                if verbose:
                    sys.stdout.write(f"\n\033[36m[Step {step_count} Start]\033[0m\n")
                    sys.stdout.flush()
            elif ev_type == "text":
                if t_first_content is None:
                    t_first_content = now
                txt = part.get("text", "")
                if txt:
                    full_content.append(txt)
                    if verbose:
                        sys.stdout.write(txt)
                        sys.stdout.flush()
            elif ev_type == "tool_use":
                if t_first_content is None:
                    t_first_content = now
                tool_name = part.get("tool") or part.get("name") or "tool"
                tools_used.append(tool_name)
                if verbose:
                    sys.stdout.write(f"\033[33m[Tool: {tool_name}]\033[0m ")
                    sys.stdout.flush()
            elif ev_type == "step_finish":
                toks = part.get("tokens", {})
                total_reasoning_tokens += toks.get("reasoning", 0)
                total_output_tokens += toks.get("output", 0)
                if verbose:
                    sys.stdout.write(f"\n\033[90m[Step {step_count} Finish: {toks.get('reasoning',0)} r-tok, {toks.get('output',0)} out-tok]\033[0m\n")
                    sys.stdout.flush()

        proc.wait()
    except Exception as e:
        print(f"\nOpenCode execution error: {e}", file=sys.stderr)
        return None

    t_end = time.time()
    total_time = max(0.001, t_end - t_start)
    startup_sec = (t_first_step - t_start) if t_first_step else 0.0
    ttft_sec = (t_first_content - t_start) if t_first_content else total_time

    total_tokens = total_reasoning_tokens + total_output_tokens
    gen_time = max(0.001, total_time - startup_sec)
    tps = (total_tokens / gen_time) if gen_time > 0 else 0.0

    return {
        "engine": "opencode",
        "model": model_arg,
        "variant": variant,
        "startup_sec": startup_sec,
        "ttft_sec": ttft_sec,
        "thinking_dur_sec": max(0.0, ttft_sec - startup_sec),
        "thinking_tokens": total_reasoning_tokens,
        "output_dur_sec": max(0.0, total_time - ttft_sec),
        "output_tokens": total_output_tokens,
        "steps": step_count,
        "tools": tools_used,
        "total_dur_sec": total_time,
        "total_tokens": total_tokens,
        "tps": tps,
        "exit_code": proc.returncode,
        "content": "".join(full_content)
    }

def run_gateway_bench(model, thinking_level="high", prompt="hi", verbose=False):
    base_url, api_key = resolve_gateway()
    if not api_key:
        print("Error: No API key found in ~/.config/opencode/opencode.json or OPENCODE_API_KEY", file=sys.stderr)
        return None

    target_model = model.replace("10router/", "")
    if target_model.startswith("ag/gemini-") and thinking_level in ("low", "medium", "high"):
        parts = target_model.split("-")
        if parts[-1] in ("low", "medium", "high"):
            parts[-1] = thinking_level
            target_model = "-".join(parts)
        else:
            target_model = f"{target_model}-{thinking_level}"

    payload = {
        "model": target_model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": True,
        "max_tokens": 1500
    }

    if thinking_level in ("low", "medium", "high"):
        payload["reasoning_effort"] = thinking_level
        payload["thinkingConfig"] = {"thinkingLevel": thinking_level}

    url = f"{base_url}/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        "User-Agent": "opencode/1.18.25"
    }

    req = urllib.request.Request(url, data=json.dumps(payload).encode("utf-8"), headers=headers)

    t_start = time.time()
    t_first_token = None
    t_first_content = None
    t_reasoning_start = None
    t_reasoning_end = None

    reasoning_text = []
    content_text = []
    reasoning_tokens_reported = None
    output_tokens_reported = None

    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            for raw_line in resp:
                line = raw_line.decode("utf-8", errors="replace").strip()
                if not line or line.startswith(":"):
                    continue
                if line.startswith("data: "):
                    data_str = line[6:]
                    if data_str == "[DONE]":
                        break
                    try:
                        chunk = json.loads(data_str)
                    except Exception:
                        continue

                    now = time.time()
                    if t_first_token is None:
                        t_first_token = now

                    choices = chunk.get("choices", [])
                    if choices:
                        delta = choices[0].get("delta", {})
                        r = delta.get("reasoning_content") or delta.get("thinking") or ""
                        c = delta.get("content") or ""

                        if r:
                            if t_reasoning_start is None:
                                t_reasoning_start = now
                            t_reasoning_end = now
                            reasoning_text.append(r)
                            if verbose:
                                sys.stdout.write(f"\033[90m{r}\033[0m")
                                sys.stdout.flush()

                        if c:
                            if t_first_content is None:
                                t_first_content = now
                            content_text.append(c)
                            if verbose:
                                sys.stdout.write(c)
                                sys.stdout.flush()

                    usage = chunk.get("usage")
                    if usage:
                        comp_details = usage.get("completion_tokens_details", {})
                        reasoning_tokens_reported = comp_details.get("reasoning_tokens")
                        output_tokens_reported = usage.get("completion_tokens")
    except Exception as e:
        print(f"\nGateway API Error: {e}", file=sys.stderr)
        return None

    t_end = time.time()
    total_time = max(0.001, t_end - t_start)
    ttft = (t_first_token - t_start) if t_first_token else total_time

    full_reasoning = "".join(reasoning_text)
    full_content = "".join(content_text)

    r_tokens = reasoning_tokens_reported if reasoning_tokens_reported is not None else max(len(full_reasoning) // 4, 1 if full_reasoning else 0)
    o_tokens = output_tokens_reported if output_tokens_reported is not None else max(len(full_content) // 4, 1 if full_content else 0)

    if full_reasoning and t_first_content:
        thinking_dur = max(0.0, t_first_content - t_first_token)
    elif full_reasoning and t_reasoning_end and t_reasoning_start:
        thinking_dur = max(0.0, t_reasoning_end - t_reasoning_start)
    else:
        thinking_dur = 0.0

    output_dur = max(0.0, t_end - (t_first_content or t_first_token or t_end))
    total_tokens = r_tokens + o_tokens
    tps = (total_tokens / total_time) if total_time > 0 else 0.0

    return {
        "engine": "gateway",
        "model": target_model,
        "variant": thinking_level,
        "startup_sec": 0.0,
        "ttft_sec": ttft,
        "thinking_dur_sec": thinking_dur,
        "thinking_tokens": r_tokens,
        "output_dur_sec": output_dur,
        "output_tokens": o_tokens,
        "steps": 1,
        "tools": [],
        "total_dur_sec": total_time,
        "total_tokens": total_tokens,
        "tps": tps,
        "exit_code": 0,
        "content": full_content
    }

def print_table(results):
    header = f"{'Engine':8s} | {'Model':30s} | {'Variant':7s} | {'Startup':7s} | {'TTFT':6s} | {'Think(s)':8s} | {'R-Tok':7s} | {'Gen(s)':6s} | {'Out-Tok':7s} | {'Steps':5s} | {'Total':6s} | {'Speed':8s}"
    sep = "-" * len(header)
    print("\n" + header)
    print(sep)
    for r in results:
        if not r: continue
        startup_str = f"{r['startup_sec']:5.2f}s" if r['startup_sec'] > 0 else "   -   "
        print(f"{r['engine']:8s} | {r['model']:30s} | {r['variant']:7s} | {startup_str:7s} | {r['ttft_sec']:5.2f}s | {r['thinking_dur_sec']:7.2f}s | {r['thinking_tokens']:7d} | {r['output_dur_sec']:5.2f}s | {r['output_tokens']:7d} | {r.get('steps',1):5d} | {r['total_dur_sec']:5.2f}s | {r['tps']:6.1f} t/s")
    print(sep + "\n")

def main():
    parser = argparse.ArgumentParser(description="OpenCode Model + Thinking Speed & Latency Benchmark")
    parser.add_argument("--engine", "-e", choices=["opencode", "gateway"], default="opencode", help="Engine to benchmark: 'opencode' (default) or 'gateway'")
    parser.add_argument("--model", "-m", default="combo-coding", help="Model name (e.g. combo-coding, 10router/ag/gemini-3.8-flash-high, cx/gpt-5.6-sol)")
    parser.add_argument("--variant", "-t", default="high", help="Model variant / thinking level (e.g. high, medium, low, max, none)")
    parser.add_argument("--preset", "-p", choices=["spec", "bug", "arch"], default="bug", help="Use preset prompt (spec, bug, arch)")
    parser.add_argument("--prompt", help="Custom prompt string")
    parser.add_argument("--compare", "-c", action="store_true", help="Run comparison matrix of popular models & thinking levels")
    parser.add_argument("--verbose", "-v", action="store_true", help="Stream response live to terminal")

    args = parser.parse_args()
    prompt = args.prompt or PRESETS[args.preset]

    if args.compare:
        print(f"=== Running Comparison Matrix [{args.engine.upper()}] on Preset '{args.preset}' ===")
        combos = [
            ("10router/ag/gemini-3.8-flash-high", "high"),
            ("10router/ag/gemini-3.8-flash-medium", "medium"),
            ("10router/ag/gemini-3.7-flash-high", "high"),
            ("10router/combo-coding", "high"),
        ]
        results = []
        for mod, vr in combos:
            print(f"Testing {mod} [{vr}]...", end="", flush=True)
            if args.engine == "opencode":
                res = run_opencode_bench(mod, vr, prompt=prompt, verbose=False)
            else:
                res = run_gateway_bench(mod, vr, prompt=prompt, verbose=False)

            if res:
                print(f" done in {res['total_dur_sec']:.2f}s ({res['tps']:.1f} tps)")
                results.append(res)
            else:
                print(" FAILED")
        print_table(results)
    else:
        print(f"Testing [{args.engine.upper()}] {args.model} [variant={args.variant}] on preset '{args.preset}'...")
        if args.engine == "opencode":
            res = run_opencode_bench(args.model, args.variant, prompt=prompt, verbose=args.verbose)
        else:
            res = run_gateway_bench(args.model, args.variant, prompt=prompt, verbose=args.verbose)

        if res:
            print_table([res])
        else:
            print("Benchmark failed.")

if __name__ == "__main__":
    main()
